src/blog/understanding-ecma-part-1.md

```diff
@@ -0,0 +1,232 @@
+
+The most well-known exotic object is the _Array_, since its length property behaves in a non-default way: setting the length property can remove elements from the _Array_.
+
+Essential internal methods are the methods listed [here](https://tc39.es/ecma262/#table-5).
+
+## Completion Records
+
+What about the question marks and exclamation marks? To understand them, we need to look into **Completion Records**!
+
+A Completion Record is a "record" (or a "struct", if you're more familiar with that term) defined by the spec. A record is a data type which has a fixed set of named fields. C-like languages have structs - JavaScript doesn't, since you can always add properties into an object.
+
+Again, the Completion Record is defined only for spec purposes. A JavaScript engine doesn't have to have a corresponding internal data type.
+
+A Completion Record has three fields:
+
+| Name | Description |
+--- | ---
+| \[\[Type\]\] | One of: normal, break, continue, return, or throw. All other types except "normal" are referred to as "abrupt completions".|
+| \[\[Value\]\] | the value that was produced when the completion occurred, e.g., the return value of a function or the exception (if one is thrown)|
+| \[\[Target\]\] | used for directed control transfers (ignoring it for this blog post)|
+
+[Spec: Completion Record](https://tc39.es/ecma262/#sec-completion-record-specification-type)
+
+Every abstract operation implicitly returns a Completion Record. Even if it looks like an abstract operation would return a simple type such as Boolean, it's implicitly wrapped into a Completion Record with the type "normal" (see [Implicit Completion Values](https://www.ecma-international.org/ecma-262/index.html#sec-implicit-completion-values)).
+
+If an algorithm throws an exception, it means returning a Completion Record with [[Type]] throw whose [[Value]] is the exception object. We'll ignore the break, continue and return types for now.
+
+_ReturnIfAbrupt(argument)_ means taking the following steps:
+
+> 1. If argument is abrupt, return argument
+> 2. Set argument to argument.\[\[Value\]\]
+
+That is, we inspect a Completion Record; if it's an abrupt completion, we return immediately. Otherwise, we extract the value from the Completion Record.
+
+[Spec: ReturnIfAbrupt](https://tc39.es/ecma262/#sec-returnifabrupt)
+
+_ReturnIfAbrupt_ might look like a function call, but it's not. It causes the function where _ReturnIfAbrupt()_ occurs to return, not the "ReturnIfAbrupt" function itself. It behaves more like a macro in C-like languages.
+
+_ReturnIfAbrupt_ can be used like this:
+
+> let obj be Foo()
+> // obj is a Completion Record
+> ReturnIfAbrupt(obj)
+> // If we're still here, obj is now the value extracted from the Completion Record
+> Bar(obj)
+
+And now the question mark comes into play: ? Foo() is equivalent to _ReturnIfAbrupt(Foo())_.
+
+Similarly, Let val be ! Foo() is equivalent to:
```

Comment:
Here and elsewhere: the specese snippets like "Let val be ! Foo()" be styled differently.